### PR TITLE
Provide semantic tokens only for standard constructors

### PR DESCRIPTION
**Pull request by mlutze**
Fixes https://github.com/flix/flix/issues/2551
Screenshot from 2021-11-02 10-10-14

Hovering is still not changed, however.

Screenshot from 2021-11-02 10-11-14

I guess that has to be done in `HoverProvider`

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala [hunk 1]
```diff
@@ -516,20 +516,69 @@ object SemanticTokensProvider {
...
+    case TypeConstructor.Char => true
+    case TypeConstructor.Float32 => true
+    case TypeConstructor.Float64 => true
+    case TypeConstructor.Int8 => true
+    case TypeConstructor.Int16 => true
+    case TypeConstructor.Int32 => true
+    case TypeConstructor.Int64 => true
+    case TypeConstructor.BigInt => true
+    case TypeConstructor.Str => true
+    case TypeConstructor.Arrow(arity) => false
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala:554]: Unused variables should be `_`.

File: main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala [hunk 2]
```diff
@@ -516,20 +516,69 @@ object SemanticTokensProvider {
...
 
     case Type.UnkindedVar(_, _, _, _) =>
       throw InternalCompilerException(s"Unexpected type: '$tpe0'.")
   }
 
+  /**
+    * Returns true if the type constructor should be highlighted.
+    * This is restricted to type constructors whose that use the standard shape (X[Y, Z]).
+    */
+  private def shouldColor(tycon: TypeConstructor): Boolean = tycon match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala:541]: Wish we had a better name. Maybe isSemanticToken? or isVisibleTypeConstructor?

**Review by magnus-madsen — COMMENTED** [review]
I am not sure why this is the right solution. For example, why is "and" not colored? Is it not because the "and" has the wrong (too wide) source location?

**Comment by mlutze** [comment]
> I am not sure why this is the right solution. For example, why is "and" not colored? Is it not because the "and" has the wrong (too wide) source location?

It might have the right location; I can test that. I chose not to highlight those because I think of them as type operators rather than type constructors.

**Comment by magnus-madsen** [comment]
> > I am not sure why this is the right solution. For example, why is "and" not colored? Is it not because the "and" has the wrong (too wide) source location?
> 
> It might have the right location; I can test that. I chose not to highlight those because I think of them as type operators rather than type constructors.

My question remains: Why is this not a source location problem, but rather a filtering problem?

**Comment by mlutze** [comment]
> My question remains: Why is this not a source location problem, but rather a filtering problem?

We could look at it as a location problem. But there are some constructors which cannot have a reasonable semantic token location, IMO. What is a record constructor's location, for example?

In general, there's not a direct correspondence between the shape of types syntactically and semantically.

**Comment by magnus-madsen** [comment]
Thanks!

**Pull request merged**